### ISSUE TITLE
Memory usage?

### ISSUE THREAD

**Issue by sdrabblescripta** [samuller/log-rate-limit#1]
This looks like a really useful tool that will solve a problem I'm having, but looking at the code it looks like the memory usage can grow unbounded. Can you tell me how/ when memory is flushed? I have a number of very large log messages (full stack traces) and don't want to exhaust the service memory with no way to clear things out.

**Comment by samuller** [comment]
Yes, that's a valid concern as the current implementation's memory usage will grow for each unique `stream_id` (by default this means each unique log message) which definitely has the potential for issues with long running processes that have lots of large and unique log messages (or even just messages that contain some small changing values, e.g. timestamps - although a custom constant `stream_id`will help for this).

I actually found I already have a partial implementation (uncommitted) of a clearing approach that would expire log messages from the cache - but it'll require some clean-up and testing.

**Comment by samuller** [comment]
`v1.3.0` now expires streams after some time by default (currently 15 minutes after the end of the rate-limit period for the last outputted log). Everything's configurable, but stream clearing can also be triggered manually by using `clear_old_streams()`.

Memory use is still unbounded in the sense that there are no limits to the length of logs as well as the number of unique logs, although the first of those can now be handled with the new `stream_id_max_len` option which can be used to limit the length of  `stream_ids`. It defaults to unlimited as it introduces the risk of confusing logs when their first `stream_id_max_len` characters are identical.

I wouldn't recommend using the new version immediately in production as I've yet to run it for a while on some of our internal systems to confirm it hasn't introduced any new issues.

**Comment by sdrabblescripta** [comment]
Thanks for jumping on this. I don't know that I'll have a chance to try out the new version any time soon but I appreciate you looking into it!
